doc/neps/nep-0050-scalar-promotion.rst

```diff
@@ -0,0 +1,729 @@
+
+     int64_0d_array = np.array(1, dtype=np.int64)
+     np.result_type(np.int8, int64_0d_array) == np.int8
+
+   This logic arises because ``1`` can be represented by an ``int8`` while
+   ``255`` can be represented by an ``int16`` *or* ``uint8``.
+
+   Because of this, the exact type is often ignored for 0-D arrays or
+   NumPy scalars.
+
+2. For a Python ``int``, ``float``, or ``complex`` the value is inspected as
+   before.  But surprisingly *not* when the NumPy object is a 0-D array
+   or NumPy scalar::
+
+     np.result_type(np.array(1, dtype=np.uint8), 1) == np.int64
+     np.result_type(np.int8(1), 1) == np.int64
+
+   The reason is that the special value-based promotion is disabled when all
+   objects are scalars or 0-D arrays.
+   NumPy thus returns the same type as ``np.array(1)``, which is usually
+   an ``int64`` (this depends on the system).
+
+Note that the examples apply also to operations like multiplication,
+addition, comparisons, and their corresponding functions like ``np.multiply``.
+
+This NEP proposes to refactor the behaviour around two guiding principles:
+
+1. Values must never influence result type.
+2. NumPy scalars or 0-D arrays should behave consistently with their
+   N-D counterparts.
+
+We propose to remove all value-based logic and add special handling for
+Python scalars to preserve some convenient behaviors.
+Python scalars will be considered "weakly" typed.
+When a NumPy array/scalar is combined with a Python scalar, it will
+be converted to the NumPy dtype, such that::
+
+    np.array([1, 2, 3], dtype=np.uint8) + 1  # returns a uint8 array
+    np.array([1, 2, 3], dtype=np.float32) + 2.  # returns a float32 array
+
+There will be no dependence on the Python value itself.
+
+The proposed changes also apply to ``np.can_cast(100, np.int8)``, however,
+we expect that the behaviour in functions (promotion) will, in practice, be far
+more important than the casting change itself.
+
+
+Schema of the new proposed promotion rules
+------------------------------------------
+
+After the change, the promotions in NumPy will follow the schema below.
+Promotion always occurs along the green lines:
+from left to right within their kind and to a higher kind only when
+necessary.
+The result kind is always the largest kind of the inputs.
+Note that NumPy considers ``float32`` to have a lower precision than ``int32``
+and ``uint32``, but makes an exception for ``float64`` (and higher).
+
+The Python scalars are inserted at the very left of each "kind" and the
+Python integer does not distinguish signed and unsigned.
+Note, that when the promoting a Python scalar with a dtype of lower kind
+category (boolean, integral, inexact) with a higher one, we  use the
+minimum/default precision: that is ``float64``, ``complex128`` or ``int64``
+(``int32`` is used on some systems, e.g. windows).
+
+.. figure:: _static/nep-0050-promotion-no-fonts.svg
+    :figclass: align-center
```

Comment:
The horizontal lines in this graphic should have the same thickness and colour as the diagonal ones. At first glance it looks like `uint8` can only promote to `float16`, not `uint16`.

![grafik](https://user-images.githubusercontent.com/33685575/163495919-987c3816-0ce0-496e-abaa-e9c9263ea13f.png)

Reply:
Ah, got some alpha wrong or so, will fix, thanks.